### ISSUE TITLE
bring nosetest job test_builtin.py back to work

### ISSUE THREAD

**Issue by tai271828** [solvcon/pastsolvcon#135]
I would like to develop parallel code for probe feature and will use this feature to study Sod tube example.
I want to know how SOLVCON run parallel jobs by studying ftests/parallel/*

Let's bring the malfunction test_builtin.py back to work first.

CHECK LIST:
issue #158

**Comment by tai271828** [comment]
First of all,
let me try to narrow down the case study to two test cases are run.
They are

```
class TestSequential(TestBlockCaseRun):
    ....
    def test_dsoln(self):
    ....
class TestLocalParallel(TestBlockCaseRun):
    npart = 3
    def test_dsoln(self):
    ....
```

test_dsoln of TestSequential could pass nosetest but not TestLocalParallel does.

and the test result is:

```
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
Process Process-3:
Traceback (most recent call last):
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
Process Process-1:
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
Traceback (most recent call last):
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 205, in run
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 205, in run
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self.eventloop()
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 159, in eventloop
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 205, in run
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 138, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 334, in march
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 159, in eventloop
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 138, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 334, in march
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 159, in eventloop
    method(worker=worker)
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 80, in calcsoln
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 138, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 334, in march
    self.create_alg().calc_soln()
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 65, in create_alg
    alg.setup_mesh(self.blk)
    method(worker=worker)
  File "mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3777)
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 80, in calcsoln
    self.create_alg().calc_soln()
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 65, in create_alg
    method(worker=worker)
    alg.setup_mesh(self.blk)
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 80, in calcsoln
  File "mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3777)
  File "mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4387)
    self.create_alg().calc_soln()
AttributeError: ndcrd array mismatch: shared(9530496,37111600), body(0,37032432)
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 65, in create_alg
    alg.setup_mesh(self.blk)
  File "mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3777)
  File "mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4387)
  File "mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4387)
AttributeError: ndcrd array mismatch: shared(9530496,37142208), body(0,37030592)
AttributeError: ndcrd array mismatch: shared(9530496,37108208), body(0,37050880)
E.
======================================================================
ERROR: test_dsoln (test_tmp.TestLocalParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 151, in test_dsoln
    case.run()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py", line 814, in run
    self._run_march()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py", line 872, in _run_march
    self.execution.marchret = [sdw.recv() for sdw in dealer]
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 430, in recv
    return self.conn.recv(*args, **kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/connection.py", line 145, in recv
    return self.conn.recv(*args, **kw)
EOFError: 
-------------------- >> begin captured stdout << ---------------------
********************************************************************************
*** Start init (level 0) blockcase ... 
*** ****************************************************************************
*** ****************************************************************************
*** *** Start build_domain ... 
*** *** ************************************************************************
*** *** mesh file: /home/tai271828/work-my-projects/solvcon/solvcon.installed/test/data/sample.neu
*** *** ************************************************************************
*** *** *** Start read_neu_data from /home/tai271828/work-my-projects/solvcon/solvcon.installed/test/data/sample.neu ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End read_neu_data . Elapsed time (sec) = 4.50611e-05
*** *** ************************************************************************
*** *** ************************************************************************
*** *** *** Start create_neu_object ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End create_neu_object . Elapsed time (sec) = 0.00236607
*** *** ************************************************************************
*** *** ************************************************************************
*** *** *** Start convert_neu_to_block ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End convert_neu_to_block . Elapsed time (sec) = 0.00325108
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End build_domain . Elapsed time (sec) = 0.00580096
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start split_domain ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End split_domain . Elapsed time (sec) = 0.00576496
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start build_dealer ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End build_dealer . Elapsed time (sec) = 0.30845
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start interconnect ... 
*** *** ************************************************************************
*** *** 0 -> 1 2.
*** *** 1 -> 2.
*** *** ************************************************************************
*** *** End interconnect . Elapsed time (sec) = 0.303812
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start remote_init_solver ... 
*** *** ************************************************************************
*** *** solver #0/(3-1): sending ... done.
*** *** solver #1/(3-1): sending ... done.
*** *** solver #2/(3-1): sending ... done.
*** *** Bind/Init ... done.
*** *** ************************************************************************
*** *** End remote_init_solver . Elapsed time (sec) = 0.0043819
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start init_interface ... 
*** *** ************************************************************************
*** *** Interface exchanging pairs (3 phases):
*** *** 0 -> 0-1 0-2 ---
*** *** 1 -> 0-1 --- 1-2
*** *** 2 -> --- 0-2 1-2
*** *** ************************************************************************
*** *** End init_interface . Elapsed time (sec) = 0.000276089
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start exchange_metric ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End exchange_metric . Elapsed time (sec) = 0.000146866
*** ****************************************************************************
*** ****************************************************************************
*** End init blockcase . Elapsed time (sec) = 0.629264
********************************************************************************
********************************************************************************
*** Start run (level 0) blockcase ... 
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start loop_march ... 
*** *** ************************************************************************
*** *** 
--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 2 tests in 0.652s

FAILED (errors=1)


```

**Comment by tai271828** [comment]
There is solver_legacy.py in the error message.
Looks like I need to port the legacy code into new one?

**Comment by tai271828** [comment]
try to play around with the variable `npart`
it looks like to mean the core number will be used in parallel mode.
Ref the dumping message

```
Process Process-2:
Traceback (most recent call last):
Process Process-1:
......

```

**Comment by tai271828** [comment]
hmmm, set `npart` as 1 will make the Traceback message more readable : D

```
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 205, in run
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 159, in eventloop
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 138, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 334, in march
    method(worker=worker)
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 80, in calcsoln
    self.create_alg().calc_soln()
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 65, in create_alg
    alg.setup_mesh(self.blk)
  File "mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3777)
  File "mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4387)
AttributeError: ndcrd array mismatch: shared(9530496,31823152), body(0,31692672)
E.
======================================================================
ERROR: test_dsoln (test_tmp.TestLocalParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 151, in test_dsoln
    case.run()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py", line 814, in run
    self._run_march()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py", line 872, in _run_march
    self.execution.marchret = [sdw.recv() for sdw in dealer]
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 430, in recv
    return self.conn.recv(*args, **kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/connection.py", line 145, in recv
    return self.conn.recv(*args, **kw)
EOFError: 
-------------------- >> begin captured stdout << ---------------------
********************************************************************************
*** Start init (level 0) blockcase ... 
*** ****************************************************************************
*** ****************************************************************************
*** *** Start build_domain ... 
*** *** ************************************************************************
*** *** mesh file: /home/tai271828/work-my-projects/solvcon/solvcon.installed/test/data/sample.neu
*** *** ************************************************************************
*** *** *** Start read_neu_data from /home/tai271828/work-my-projects/solvcon/solvcon.installed/test/data/sample.neu ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End read_neu_data . Elapsed time (sec) = 4.81606e-05
*** *** ************************************************************************
*** *** ************************************************************************
*** *** *** Start create_neu_object ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End create_neu_object . Elapsed time (sec) = 0.00248003
*** *** ************************************************************************
*** *** ************************************************************************
*** *** *** Start convert_neu_to_block ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End convert_neu_to_block . Elapsed time (sec) = 0.00324202
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End build_domain . Elapsed time (sec) = 0.005898
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start split_domain ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End split_domain . Elapsed time (sec) = 0.00199509
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start build_dealer ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End build_dealer . Elapsed time (sec) = 0.102372
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start interconnect ... 
*** *** ************************************************************************
*** *** .
*** *** ************************************************************************
*** *** End interconnect . Elapsed time (sec) = 0.000490904
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start remote_init_solver ... 
*** *** ************************************************************************
*** *** solver #0/(1-1): sending ... done.
*** *** Bind/Init ... done.
*** *** ************************************************************************
*** *** End remote_init_solver . Elapsed time (sec) = 0.00489593
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start init_interface ... 
*** *** ************************************************************************
*** *** Interface exchanging pairs (0 phases):
*** *** 0 ->
*** *** ************************************************************************
*** *** End init_interface . Elapsed time (sec) = 0.000218153
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start exchange_metric ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End exchange_metric . Elapsed time (sec) = 0.000170946
*** ****************************************************************************
*** ****************************************************************************
*** End init blockcase . Elapsed time (sec) = 0.116398
********************************************************************************
********************************************************************************
*** Start run (level 0) blockcase ... 
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start loop_march ... 
*** *** ************************************************************************
*** *** 
--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 2 tests in 0.138s

FAILED (errors=1)

```

**Comment by tai271828** [comment]
solvcon/solver.py shows

```
 48 # import legacy.
 49 from .solver_legacy import (
 50     BaseSolverExedata, BaseSolver, FakeBlockVtk, BlockSolver)
```

so sad. looks like I need to implement a new BlockSolver?? : p

**Comment by tai271828** [comment]
not really?
let's go back to look into `ndcrd` array of mesh.pyx in more details : )

**Comment by tai271828** [comment]
tried to use ipython wih latest source code to reproduce the failure interactively
and test_soln failed because

```
<class 'list'>
Process Process-1:
Traceback (most recent call last):
  File "/home/tai271828/miniconda/lib/python3.5/multiprocessing/process.py", line 254, in _bootstrap
    self.run()
  File "/home/tai271828/miniconda/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 209, in run
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 163, in eventloop
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 142, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 629, in init_exchange
    if pair < 0:
TypeError: unorderable types: list() < int()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-a9fd3d818a0e> in <module>()
----> 1 aa.test_soln()

/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py in test_soln(self)
    163         from numpy import zeros
    164         from solvcon.domain import Collective
--> 165         case = self._get_case(npart=self.npart, domaintype=Collective)
    166         case.run()
    167         # get result.

/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py in _get_case(self, **kw)
    118         )
    119         case.runhooks.append(CaseCollect)
--> 120         case.init()
    121         return case
    122 

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py in init(self, level)
    491             self.info('\n')
    492             self._log_start('init_interface')
--> 493             self._init_interface()
    494             self._log_end('init_interface')
    495             # initialize exchange for remote solver objects.

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py in _init_interface(self)
    769             self.info(('%%0%dd ->' % dwidth) % iblk)
    770             for pair in ifacelist:
--> 771                 if pair < 0:
    772                     stab = '-' * (2*dwidth+1)
    773                 else:

TypeError: unorderable types: list() < int()

In [4]: 


```

this failure is different from the previous result : (

**Comment by tai271828** [comment]
set `npart=1` the failure message is dfferent.
Maybe I should not test the code on different local machines?

```
Traceback (most recent call last):
  File "/home/tai271828/miniconda/lib/python3.5/multiprocessing/process.py", line 254, in _bootstrap
    self.run()
  File "/home/tai271828/miniconda/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 209, in run
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 163, in eventloop
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 142, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 339, in march
    method(worker=worker)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py", line 79, in calcsoln
    self.create_alg().calc_soln()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py", line 64, in create_alg
    alg.setup_mesh(self.blk)
  File "solvcon/mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3870)
  File "solvcon/mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4507)
AttributeError: ndcrd array mismatch: shared(5,37241264), body(0,37154560)
---------------------------------------------------------------------------
EOFError                                  Traceback (most recent call last)
<ipython-input-3-a9fd3d818a0e> in <module>()
----> 1 aa.test_soln()

/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py in test_soln(self)
    164         from solvcon.domain import Collective
    165         case = self._get_case(npart=self.npart, domaintype=Collective)
--> 166         case.run()
    167         # get result.
    168         soln = case.execution.var['soln'][:,0]

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py in run(self, level)
    814             self._run_preloop()
    815         if level < 2:
--> 816             self._run_march()
    817             self._run_postloop()
    818             self._run_exhaust()

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py in _run_march(self)
    872                     time_current, time_increment, steps_stride,
    873                     with_worker=True)
--> 874                 self.execution.marchret = [sdw.recv() for sdw in dealer]
    875             else:
    876                 self.execution.marchret = self.solver.solverobj.march(

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py in <listcomp>(.0)
    872                     time_current, time_increment, steps_stride,
    873                     with_worker=True)
--> 874                 self.execution.marchret = [sdw.recv() for sdw in dealer]
    875             else:
    876                 self.execution.marchret = self.solver.solverobj.march(

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py in recv(self, *args, **kw)
    432         Receive data from worker/muscle.
    433         """
--> 434         return self.conn.recv(*args, **kw)
    435 
    436 class Dealer(list):

/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/connection.py in recv(self, *args, **kw)
    150         return self.conn.send(*args, **kw)
    151     def recv(self, *args, **kw):
--> 152         return self.conn.recv(*args, **kw)
    153     def close(self, *args, **kw):
    154         return self.conn.close(*args, **kw)

/home/tai271828/miniconda/lib/python3.5/multiprocessing/connection.py in recv(self)
    248         self._check_closed()
    249         self._check_readable()
--> 250         buf = self._recv_bytes()
    251         return ForkingPickler.loads(buf.getbuffer())
    252 

/home/tai271828/miniconda/lib/python3.5/multiprocessing/connection.py in _recv_bytes(self, maxsize)
    405 
    406     def _recv_bytes(self, maxsize=None):
--> 407         buf = self._recv(4)
    408         size, = struct.unpack("!i", buf.getvalue())
    409         if maxsize is not None and size > maxsize:

/home/tai271828/miniconda/lib/python3.5/multiprocessing/connection.py in _recv(self, size, read)
    381             if n == 0:
    382                 if remaining == size:
--> 383                     raise EOFError
    384                 else:
    385                     raise OSError("got end of file during message")

EOFError:
```

**Comment by yungyuc** [comment]
npart=1 could be a corner case for the code.  It's not useful and I didn't test that case a lot when the code was still alive.  I think for code quality it should be made working, but in my humble opinion it's not as important as other failures.

I think testing the parallel code on different hosts should be made passing.  It represents normal use cases.

When seeing the error message in the communication code, it might be helpful to add debugging code to print what's really in the message buffer.

**Comment by tai271828** [comment]
tried npart=2

```
└──╼ nosetests test_builtin.py
Process Process-1:
E
======================================================================
ERROR: test_soln (test_builtin.TestLocalParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py", line 165, in test_soln
    case = self._get_case(npart=self.npart, domaintype=Collective)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.160226.parallel/ftests/parallel/test_builtin.py", line 120, in _get_case
    case.init()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py", line 493, in init
    self._init_interface()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/case_legacy.py", line 771, in _init_interface
    if pair < 0:
TypeError: unorderable types: list() < int()
-------------------- >> begin captured stdout << ---------------------
********************************************************************************
*** Start init (level 0) blockcase ... 
*** ****************************************************************************
*** ****************************************************************************
*** *** Start build_domain ... 
*** *** ************************************************************************
*** *** mesh file: /home/tai271828/work-my-projects/solvcon/solvcon.installed/test/data/sample.neu
*** *** ************************************************************************
*** *** *** Start read_neu_data from /home/tai271828/work-my-projects/solvcon/solvcon.installed/test/data/sample.neu ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End read_neu_data . Elapsed time (sec) = 0.0119057
*** *** ************************************************************************
*** *** ************************************************************************
*** *** *** Start create_neu_object ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End create_neu_object . Elapsed time (sec) = 0.00262833
*** *** ************************************************************************
*** *** ************************************************************************
*** *** *** Start convert_neu_to_block ... 
*** *** *** ********************************************************************
*** *** *** ********************************************************************
*** *** *** End convert_neu_to_block . Elapsed time (sec) = 0.00545597
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End build_domain . Elapsed time (sec) = 0.0281312
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start split_domain ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End split_domain . Elapsed time (sec) = 0.00649524
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start build_dealer ... 
*** *** ************************************************************************
*** *** ************************************************************************
*** *** End build_dealer . Elapsed time (sec) = 0.250755
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start interconnect ... 
*** *** ************************************************************************
*** *** 0 -> 1.
*** *** ************************************************************************
*** *** End interconnect . Elapsed time (sec) = 0.102115
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start remote_init_solver ... 
*** *** ************************************************************************
*** *** solver #0/(2-1): sending ... done.
*** *** solver #1/(2-1): sending ... done.
*** *** Bind/Init ... done.
*** *** ************************************************************************
*** *** End remote_init_solver . Elapsed time (sec) = 0.0167089
*** ****************************************************************************
*** 
*** ****************************************************************************
*** *** Start init_interface ... 
*** *** ************************************************************************
*** *** Interface exchanging pairs (1 phases):
*** *** 0 ->
--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.421s

FAILED (errors=1)
Traceback (most recent call last):
  File "/home/tai271828/miniconda/lib/python3.5/multiprocessing/process.py", line 254, in _bootstrap
    self.run()
  File "/home/tai271828/miniconda/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 209, in run
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 163, in eventloop
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/rpc.py", line 142, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/solvcon.installed/solvcon/solver_legacy.py", line 629, in init_exchange
    if pair < 0:
TypeError: unorderable types: list() < int()

```

**Comment by yungyuc** [comment]
@tai271828 what do you find?  Printing may be the only tool available for now.  I guess you might want to build some debugging code to help you.

**Comment by tai271828** [comment]
error message is back to (or say similar to) https://github.com/solvcon/solvcon/issues/135#issuecomment-167126426

with PR#149

**Comment by yungyuc** [comment]
Is this issue really about test_builtin.py?  In my computer it doesn't fail.  If the related changes of this issue is no longer related to test_builtin.py, I suggest to create a new issue to track the real work, and close this one.

**Comment by tai271828** [comment]
yes it is.
The following is the script how I played around with:

```
import test_builtin
test_instance = test_builtin.TestLocalParallel()
test_instance.test_dsoln()
```

**Comment by yungyuc** [comment]
My mistake.  I mistook the name "test_builtin.py" to a similar test file.  You are right it's a parallel test file.

**Comment by tai271828** [comment]
this kind of error message of https://github.com/solvcon/solvcon/issues/135#issuecomment-167126426

```
AttributeError: ndcrd array mismatch: shared(9530496,37111600), body(0,37032432)
  File "/home/tai271828/work-my-projs/solvcon/solvcon.dev.sodtube.150906/ftests/parallel/test_tmp.py", line 65, in create_alg
    alg.setup_mesh(self.blk)
  File "mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3777)
  File "mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4387)
  File "mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4387)
AttributeError: ndcrd array mismatch: shared(9530496,37142208), body(0,37030592)
AttributeError: ndcrd array mismatch: shared(9530496,37108208), body(0,37050880)
```

and what I reproduced today:

```
*** *** Start loop_march ... 
*** *** ************************************************************************
*** *** Process Process-1:
Traceback (most recent call last):
  File "/home/tai271828/anaconda/lib/python3.5/multiprocessing/process.py", line 254, in _bootstrap
    self.run()
  File "/home/tai271828/anaconda/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tai271828/work-my-projects/solvcon/16w9.fftest.build_in.soln.fix/solvcon/rpc.py", line 209, in run
    self.eventloop()
  File "/home/tai271828/work-my-projects/solvcon/16w9.fftest.build_in.soln.fix/solvcon/rpc.py", line 163, in eventloop
    self._eventloop()
  File "/home/tai271828/work-my-projects/solvcon/16w9.fftest.build_in.soln.fix/solvcon/rpc.py", line 142, in _eventloop
    ret = method(*ntc.args, **ntc.kw)
  File "/home/tai271828/work-my-projects/solvcon/16w9.fftest.build_in.soln.fix/solvcon/solver_legacy.py", line 339, in march
    method(worker=worker)
  File "/home/tai271828/work-my-projects/solvcon/16w9.fftest.build_in.soln.fix/ftests/parallel/test_builtin.py", line 80, in calcsoln
    self.create_alg().calc_soln()
  File "/home/tai271828/work-my-projects/solvcon/16w9.fftest.build_in.soln.fix/ftests/parallel/test_builtin.py", line 65, in create_alg
    alg.setup_mesh(self.blk)
  File "solvcon/mesh.pyx", line 179, in solvcon.mesh.Mesh.setup_mesh (solvcon/mesh.c:3870)
  File "solvcon/mesh.pyx", line 225, in solvcon.mesh.Mesh._check_pointers (solvcon/mesh.c:4507)
AttributeError: ndcrd array mismatch: shared(5,24195744), body(0,24380512)
```

is caused by issuing this line (in anaconda/lib/python3.5/multiprocessing/connection.py):  

```
      self._send_bytes(ForkingPickler.dumps(obj))
```

when SOLVCON try to run the march (in solvcon/case_legacy.py)

```
            if flag_parallel:
                for sdw in dealer: sdw.cmd.march(
                    time_current, time_increment, steps_stride,
                    with_worker=True)
```

**Comment by yungyuc** [comment]
@tai271828 There probably is something wrong with slicing.  I am not sure.  What's your specific question?

**Comment by tai271828** [comment]
Addresses or data of table instances seem not to pass to worker of sub-processes correctly. For example, table._nda is not None before sending it via an Agent, but it will be None after receiving by Worker. 

The steps to identify this issue:
1. fetch the code. checkout to the version by `git checkout f47e5c10806b35cdd9313c74e403f4ee160e8fed`
2. apply the patch [0001-used-for-debug-thread.patch.txt](https://github.com/solvcon/solvcon/files/380327/0001-used-for-debug-thread.patch.txt)
3. build the source `python setup.py build_ext --inplace`
4. go to `ftests/parallel` and run `nosetests  test_builtin:TestLocalParallel.test_soln -s`. please note the python version is 2.7
5. compare the `stdout` message and the debug message in `/tmp/solvcon-worker-dbg/`

**Comment by tai271828** [comment]
This issue was fixed by the commit https://github.com/solvcon/solvcon/pull/168.

Besides, the latest source code https://github.com/solvcon/solvcon/commit/b11b5a52920634e1b40f51f5642b30d755258058 shows

```
└──╼ nosetests ftests/parallel/test_builtin.py 
.......
----------------------------------------------------------------------
Ran 7 tests in 3.985s

OK
```

So please let me close this issue.
